src/transformers/trainer_utils.py

```diff
@@ -1,6 +1,83 @@
+import logging
+import os
 from typing import Dict, NamedTuple, Optional
 
 import numpy as np
+from tqdm.auto import tqdm
+
+
+logger = logging.getLogger(__name__)
+
+try:
+    import wandb
+
+    wandb.ensure_configured()
+    if wandb.api.api_key is None:
+        _has_wandb = False
+        wandb.termwarn("W&B installed but not logged in.  Run `wandb login` or set the WANDB_API_KEY env variable.")
+    else:
+        _has_wandb = False if os.getenv("WANDB_DISABLED") else True
+except ImportError:
+    _has_wandb = False
+    logger.info(
+        "You are instantiating a Trainer but W&B is not installed. To use wandb logging, "
+        "run `pip install wandb; wandb login` see https://docs.wandb.com/huggingface."
+    )
+
+
+def is_wandb_available():
+    return _has_wandb
+
+
+def setup_wandb(trainer):
+    """
+    Setup the optional Weights & Biases (`wandb`) integration.
+
+    One can override this method to customize the setup if needed.  Find more information at https://docs.wandb.com/huggingface
+    You can also override the following environment variables:
+
+    Environment:
+        WANDB_WATCH:
+            (Optional, ["gradients", "all", "false"]) "gradients" by default, set to "false" to disable gradient logging
+            or "all" to log gradients and parameters
+        WANDB_PROJECT:
+            (Optional): str - "huggingface" by default, set this to a custom string to store results in a different project
+        WANDB_DISABLED:
+            (Optional): boolean - defaults to false, set to "true" to disable wandb entirely
+    """
+    logger.info('Automatic Weights & Biases logging enabled, to disable set os.environ["WANDB_DISABLED"] = "true"')
+
+    wandb.init(project=os.getenv("WANDB_PROJECT", "huggingface"), config=vars(trainer.args))
+    # keep track of model topology and gradients (only with Pytorch)
+    if os.getenv("WANDB_WATCH") != "false" and trainer.__class__.__name__ == "Trainer":
+        wandb.watch(
+            trainer.model, log=os.getenv("WANDB_WATCH", "gradients"), log_freq=max(100, trainer.args.logging_steps)
+        )
+    # give access to wandb module
+    trainer._wandb = wandb
+
+
+def log_metrics(trainer, logs: Dict[str, float], iterator: Optional[tqdm] = None) -> None:
```

Comment:
Epoch and step might be passed as arguments which doesn't make too changes in the Trainer classes.

Reply:
Here it was mainly to keep it the same way as `Trainer`. There are a few advantages in doing it this way:

* some example scripts call for evaluation after full training (ex: `run_glue`), where the epoch & step are not accessible any longer, which were the reasons why we added them within the class.
* access to `trainer.tb_log` discussed in your previous comment will also be easier this way

Reply:
For evaluation I don't think you need the epoch. For the step yes indeed. But try to do with the `trainer.tb_log` it should be easier for me to see what you mean :)